cpp/arcticdb/version/version_core.cpp

```diff
@@ -963,27 +928,25 @@ VersionedItem compact_incomplete_impl(
 
     std::vector<folly::Future<VariantKey>> fut_vec;
     std::vector<FrameSlice> slices;
-    auto index = index_type_from_descriptor(first_seg.descriptor());
-
-    util::variant_match(index, [
-        &fut_vec, &slices, sparsify, pipeline_context=pipeline_context, &store, convert_int_to_float] (auto idx) {
-        using IndexType = decltype(idx);
-
-        if(sparsify) {
-            do_compact<IndexType, DynamicSchema, RowCountSegmentPolicy, SparseColumnPolicy>(
+    bool dynamic_schema = write_options.dynamic_schema;
+    auto policies = std::make_tuple(index_type_from_descriptor(first_seg.descriptor()), 
+                                    dynamic_schema ? VariantSchema{DynamicSchema::default_schema()} : VariantSchema{FixedSchema::default_schema()}, 
+                                    sparsify ? VariantColumnPolicy{SparseColumnPolicy{}} : VariantColumnPolicy{DenseColumnPolicy{}}
+                                    );
+    util::variant_match(std::move(policies), [
```

Comment:
This is a lot of indirection just to call `do_compact()` with different values known only at run time.

I wonder, then, if having these as template parameters is worth it.

Reply:
I don't have another choice or I need to revamp quite a lot of other classes, e.g. `SegmentAggregator` and `Aggregator`, which `do_compact()` calls.

Reply:
This is the only game in town, we need to somehow move from multiple run-time values to the associated types. It might be easier on the eye if it created the objects as l-values with names and then moved them into the tuple, rather than quite a lot of ternary operator